File: backend/providers/coverage.ts
/**
 * Coverage accounting (schema v2.3 — docs/PLAN-kapsama-sozlesmesi.md, Katman A).
 *
 * The one failure this project could never see: a mark the student made that
 * never became a card. A wrong card is visible — it carries `lowConfidence`, it
 * has a "Gözden geçir" bucket and a second-opinion button, and reading it once
 * is enough to notice. A missing card has none of that. It has no flag because
 * it has no card, and the page photo that could prove it is deleted from the
 * server after 60 days.
 *
 * Three prompt versions tried to fix this by asking (v2.6's counting step,
 * v2.7's three rounds of binding language) and none of them could be *measured*,
 * because nothing produced a number. This module is the measurement. The model
 * now writes down what it saw (`marks`) and which mark each card came from
 * (`markId`); the difference between those two lists is computed here, in plain
 * deterministic code, and no longer depends on the prompt's good intentions.
 * Judgement stays with the model, bookkeeping moves to the code (§0.8's spirit).
 *
 * Two lists come out, and both were invisible before:
 *   - `uncovered` — marks with no surviving card. The signal.
 *   - `unmarkedCardIds` — cards bound to no mark at all, which is prompt rule
 *     1's own violation ("işaretlenmemiş metin kart kaynağı değildir").
 *
 * Nothing here fails a job. A model that ignores the new fields produces
 * `reported: false` and the page behaves exactly as it did before v2.3 —
 * the same rule `sanitizeTopics` follows, for the same reason: a classification
 * nicety must never cost a paid capture.
 */

import { MARK_KINDS, type Card, type LlmOutput, type Mark, type MarkKind } from "../schemas/llmOutputTypes.js";

/**
 * How valuable a skipped mark is, low number first.
 *
 * Straight from prompt rule 3's ladder — handwriting, then the symbol tier,
 * then underline, then highlighter — so "en değerli atlanan işaret" is the
 * first row the owner sees rather than whatever happened to be topmost on the
 * page. `MARK_KINDS`'s own order is the single source of that ranking; an
 * unknown kind sorts last instead of throwing.
 */
export function markPriority(kind: string): number {
  const index = (MARK_KINDS as readonly string[]).indexOf(kind);
  return index < 0 ? MARK_KINDS.length : index;
}

export interface CoverageReport {
  /**
   * Whether the model reported a register at all.
   *
   * `false` means "no information", which is emphatically not the same as
   * "nothing uncovered": an older deployment, a model that ignored the field,
   * or a v2.2 payload all land here, and an empty `uncovered` list would
   * otherwise read as a clean bill of health.
   */
  reported: boolean;
  /** Every mark the model reported, sanitized and deduplicated. */
  marks: Mark[];
  /** Marks no surviving card claims, most valuable tier first. */
  uncovered: Mark[];
  /**
   * Ids of surviving cards bound to no mark. Ids only — card text is never
   * copied into a report that ends up in a log line (§7.3).
   */
  unmarkedCardIds: string[];
}

/** A mark with a usable id, a known tier and a non-empty quote, or `null`. */
function cleanMark(value: unknown): Mark | null {
  if (typeof value !== "object" || value === null || Array.isArray(value)) return null;
  const record = value as { id?: unknown; kind?: unknown; quote?: unknown };
  if (typeof record.id !== "string" || !record.id.trim()) return null;
  if (typeof record.quote !== "string" || !record.quote.trim()) return null;
  if (typeof record.kind !== "string" || !(MARK_KINDS as readonly string[]).includes(record.kind)) {
    return null;
  }
  return { id: record.id.trim(), kind: record.kind as MarkKind, quote: record.quote.trim() };
}

/**
 * Forces the v2.3 fields into a shape the rest of the server can trust, in
 * place, before validation — exactly where `sanitizeTopics` sits and for the
 * same reason.
 *
 * Drops malformed or duplicate-id marks and nulls any `markId` that does not
 * point at a surviving mark. A dangling reference is the one thing that would
 * make the whole layer lie: the card looks bound, so its mark is counted as
 * covered, so a skipped mark is reported as handled. Silent narrowing of the
 * signal is worse than the missing field, so it is resolved to `null` — which
 * shows up honestly as an unmarked card.
 */
export function sanitizeMarks(output: Record<string, unknown>): void {
  const rawMarks = output.marks;
  if (rawMarks !== undefined && !Array.isArray(rawMarks)) {
    // Not an array at all: the field is unusable, and leaving it would fail
    // schema validation for the whole page over an audit extra.
    delete output.marks;
  }

  let ids: Set<string> | null = null;
  if (Array.isArray(output.marks)) {
    const seen = new Set<string>();
    const marks: Mark[] = [];
    for (const candidate of output.marks) {
      const mark = cleanMark(candidate);
      if (!mark || seen.has(mark.id)) continue;
      seen.add(mark.id);
      marks.push(mark);
    }
    output.marks = marks;
    ids = seen;
  }

  if (!Array.isArray(output.cards)) return;
  for (const candidate of output.cards) {
    if (typeof candidate !== "object" || candidate === null || Array.isArray(candidate)) continue;
    const card = candidate as { markId?: unknown };
    if (card.markId === undefined || card.markId === null) continue;
    if (typeof card.markId !== "string" || !ids || !ids.has(card.markId.trim())) {
      card.markId = null;
      continue;
    }
    card.markId = card.markId.trim();
  }
}

export interface CoverageOptions {
  /**
   * Cards the gate rejected (the surplus over the per-page cap, a broken
   * card). They never reach the deck, so a mark they were the only claimant of
   * is *not* covered — counting them would hide precisely the case the card
   * ceiling creates, which is the one Tur A found on 18 of 18 pages.
   */
  rejectedCardIds?: readonly string[];
}

/**
 * Compares the model's mark register against the cards that will actually be
 * stored.
 *
 * Takes the two fields it needs rather than the whole output, so a caller
 * cannot accidentally make this depend on `usage` or `readText` and a test can
 * build the input by hand.
 */
export function deriveCoverage(
  output: Pick<LlmOutput, "cards"> & { marks?: readonly Mark[] },
  options: CoverageOptions = {},
): CoverageReport {
  const marks = [...(output.marks ?? [])];
  const rejected = new Set(options.rejectedCardIds ?? []);
  const surviving: Card[] = output.cards.filter((card) => !rejected.has(card.id));

  const claimed = new Set<string>();
  const unmarkedCardIds: string[] = [];
  for (const card of surviving) {
    const markId = typeof card.markId === "string" ? card.markId.trim() : "";
    if (markId) {
      claimed.add(markId);
    } else {
      unmarkedCardIds.push(card.id);
    }
  }

  const uncovered = marks
    .filter((mark) => !claimed.has(mark.id))
    // `map`+`sort`+`map` rather than a bare `sort`: `Array.prototype.sort` is
    // only guaranteed stable by index, and two marks of the same tier must keep
    // the order the model listed them in (roughly page order) instead of an
    // arbitrary one.
    .map((mark, index) => ({ mark, index }))
    .sort((a, b) => {
      const priority = markPriority(a.mark.kind) - markPriority(b.mark.kind);
      return priority !== 0 ? priority : a.index - b.index;
    })
    .map((entry) => entry.mark);

  return {
    reported: output.marks !== undefined,
    marks,
    uncovered,
    unmarkedCardIds,
  };
}

/**
 * The same derivation with "which cards survive" answered by the gate.
 *
 * Exists so the two endpoints that produce a result body (`/api/cards-vision`
 * and the `/api/jobs` worker) cannot answer that question differently. They
 * already build `output`/`gate` in lockstep; a second copy of "reject means it
 * does not count as coverage" is precisely the drift this repo keeps a
 * discipline against.
 */
export function coverageFromGate(
  output: Pick<LlmOutput, "cards"> & { marks?: readonly Mark[] },
  gate: { verdicts: ReadonlyArray<{ cardId: string; decision: string }> },
): CoverageReport {
  return deriveCoverage(output, {
    rejectedCardIds: gate.verdicts
      .filter((verdict) => verdict.decision === "reject")
      .map((verdict) => verdict.cardId),
  });
}

File: ios/CizgiCore/Sources/CizgiCore/Models/Coverage.swift
import Foundation

/// Coverage findings for one page (docs/PLAN-kapsama-sozlesmesi.md).
///
/// The failure this file exists for is the one the app could never see: a mark
/// the owner made that never became a card. A wrong card is visible — it is
/// flagged, listed under "Gözden geçir" and readable in Tekrar — while a
/// missing card has no flag, because it has no card. The page photo that could
/// prove it is the only evidence, and it is not kept forever.
///
/// Two readers answer that question and neither can answer all of it:
///
/// - **The generator's own register** (schema v2.3, `source == .generator`).
///   The model writes down every mark it saw and which mark each card came
///   from; the server subtracts one list from the other. This catches "read it
///   and did not card it" — the defect the owner reported on real pages, where
///   the starred passage sat in `readText` while the cards came from unmarked
///   text. It cannot catch a mark the model never perceived: a register does
///   not contain what its author never saw.
/// - **The independent audit** (`source == .auditor`), a different model family
///   re-reading the same photo. That is the only thing that can catch the
///   never-perceived mark, and the only reason it can is that it does not share
///   the first reader's blind spots.
///
/// Everything here is Foundation-only and pure so it can be tested by
/// `swift test` on any machine; SwiftData sees it as one JSON string on
/// `CapturedPage.coverageJSON`.
public struct PageMark: Codable, Equatable, Sendable, Identifiable {
    /// Which reader reported this mark.
    public enum Source: String, Codable, Sendable {
        /// The card generator's own register (schema v2.3).
        case generator
        /// The independent second reader (`/api/coverage`).
        case auditor
    }

    /// Stable within a page, derived from the mark's own text rather than from
    /// the model's `m1`-style label.
    ///
    /// The label is not usable as identity: it is only unique inside one
    /// response, so a page regenerated later reuses `m1` for a different
    /// passage — and a dismissal recorded against `m1` would then silently hide
    /// an unrelated mark. Deriving it from the folded quote makes "the owner
    /// dismissed this passage" mean what it says, and makes the two readers'
    /// reports comparable without either of them agreeing on labels.
    public let id: String
    public let kind: MarkKind
    /// The marked text as the reader transcribed it, verbatim from the page.
    public let quote: String
    public let source: Source

    public init(kind: MarkKind, quote: String, source: Source) {
        self.kind = kind
        self.quote = quote
        self.source = source
        self.id = PageMark.identity(kind: kind, quote: quote)
    }

    /// `kind:folded-quote`, so the same passage reported by both readers — or
    /// by the same reader twice — collapses to one identity.
    ///
    /// Folded with `CardSearch.fold` rather than `lowercased()`: on a Turkish
    /// device `"İ".lowercased()` is `"i̇"` and `"I".lowercased()` is `"ı"`, so
    /// two readings of the same word would produce two identities and the
    /// dismissal would not stick (the bug ADR-001 records, and the one that
    /// made Bilgilerim's search miss `İnflamasyon`).
    static func identity(kind: MarkKind, quote: String) -> String {
        "\(kind.rawValue):\(CardSearch.fold(quote.trimmingCharacters(in: .whitespacesAndNewlines)))"
    }

    /// Where this mark sits on prompt rule 3's ladder — handwriting first,
    /// highlighter last. Taken from `MarkKind`'s declaration order, which the
    /// contract-sync tests pin to the schema's own enum order.
    var priority: Int {
        MarkKind.allCases.firstIndex(of: kind) ?? MarkKind.allCases.count
    }
}

/// What one page's coverage looks like right now.
///
/// Stored on the page rather than derived on the fly: the generator's half
/// arrives once, with the cards, and the server's copy of the page (and of the
/// result) is deleted long before anyone reviews it.
public struct PageCoverage: Codable, Equatable, Sendable {
    /// One completed run of the independent audit.
    public struct Audit: Codable, Equatable, Sendable {
        public var performedAt: Date
        /// Marks the auditor found no card for.
        public var uncovered: [PageMark]
        /// How many marks it reported in total — the denominator for "it saw
        /// N marks and N-k of them were carded".
        public var markCount: Int
        /// Rows the server dropped as unusable. Surfaced rather than hidden: a
        /// number that climbs says the audit is confused, which is worth
        /// knowing before its findings are trusted.
        public var discarded: Int

        public init(performedAt: Date, uncovered: [PageMark], markCount: Int, discarded: Int) {
            self.performedAt = performedAt
            self.uncovered = uncovered
            self.markCount = markCount
            self.discarded = discarded
        }
    }

    /// Whether the generator reported a register at all.
    ///
    /// `false` is "no information", which is emphatically not "nothing was
    /// missed": a page captured before schema v2.3, or a model that ignored the
    /// field, lands here. Without this flag an empty list would read as a clean
    /// bill of health for a page nobody ever checked.
    public var reported: Bool
    /// Marks the generator registered and no surviving card claimed.
    public var uncovered: [PageMark]
    /// Cards the generator bound to no mark — prompt rule 1's own violation.
    /// Ids only; the cards themselves live in SwiftData.
    public var unmarkedCardIds: [String]
    /// The independent audit, once it has been run for this page.
    public var audit: Audit?
    /// Marks the owner has dismissed. Kept for ever, and kept as identities
    /// rather than as a count: a re-audit reports the same page again, and a
    /// dismissal that did not survive it would make the button useless.
    public var dismissedMarkIds: [String]

    public init(
        reported: Bool = false,
        uncovered: [PageMark] = [],
        unmarkedCardIds: [String] = [],
        audit: Audit? = nil,
        dismissedMarkIds: [String] = []
    ) {
        self.reported = reported
        self.uncovered = uncovered
        self.unmarkedCardIds = unmarkedCardIds
        self.audit = audit
        self.dismissedMarkIds = dismissedMarkIds
    }

    /// The rows to show the owner: both readers' findings, dismissals removed,
    /// duplicates collapsed, most valuable tier first.
    ///
    /// The generator's version of a mark wins a tie because it is the reading
    /// the cards were actually built from; the auditor's copy of the same
    /// passage would say the same thing in slightly different words and take a
    /// second row for it.
    public var openFindings: [PageMark] {
        // Dismissals are matched by the same overlap rule the deduplication
        // uses, not by id alone. Ids differ whenever the two readers transcribe
        // a passage slightly differently — or file it under different tiers —
        // and matching only on equality left a hole with teeth: dismissing the
        // generator's row skipped it before it could seed `seen`, so the
        // auditor's near-identical row surfaced and the owner had to dismiss
        // the same passage twice (Codex, PR #47). "Yoksay" is a decision about
        // the passage, so it has to be recognised however it is worded.
        let dismissedQuotes = dismissedMarkIds.map(PageCoverage.quote(inMarkId:))
        var seen: [String] = []
        var findings: [PageMark] = []

        for mark in uncovered + (audit?.uncovered ?? []) {
            let folded = CardSearch.fold(mark.quote)
            guard !dismissedQuotes.contains(where: { PageCoverage.overlaps($0, folded) }) else { continue }
            // Containment, not equality: two readers rarely transcribe the same
            // handwriting identically, and one of them quoting a few words more
            // does not make it a second mark. Equality here would show the
            // owner the same passage twice and cost two dismissals.
            guard !seen.contains(where: { PageCoverage.overlaps($0, folded) }) else { continue }
            seen.append(folded)
            findings.append(mark)
        }

        return findings.enumerated()
            .sorted { left, right in
                if left.element.priority != right.element.priority {
                    return left.element.priority < right.element.priority
                }
                // Stable within a tier: the readers list marks roughly in page
                // order, and re-sorting equals would hand back an order nothing
                // produced.
                return left.offset < right.offset
            }
            .map(\.element)
    }

    /// Whether either reader has anything open for this page.
    public var hasOpenFindings: Bool { !openFindings.isEmpty }

    /// Marks one finding as handled. Idempotent, so a double tap is harmless.
    ///
    /// Both ways of finishing with a mark land here: "Yoksay" (no card needed)
    /// and writing the card it was asking for. One list rather than two,
    /// because the list answers one question — does this mark still need a
    /// card? — and a second store for "resolved" would be a second place for
    /// that answer to fall out of step.
    public mutating func dismiss(_ mark: PageMark) {
        guard !dismissedMarkIds.contains(mark.id) else { return }
        dismissedMarkIds.append(mark.id)
    }

    /// Records an audit run, replacing any earlier one.
    ///
    /// Replacing rather than merging: an audit is a snapshot of "what is
    /// missing *now*", and cards added since the last run are exactly what
    /// should shrink it. Keeping both would make the list grow with every run.
    /// Dismissals are unaffected — they live outside this and are matched by
    /// identity.
    public mutating func record(audit: Audit) {
        self.audit = audit
    }

    /// The folded quote inside a mark id (`kind:folded-quote`).
    ///
    /// Reading it back out of the id rather than storing quotes separately: the
    /// id already carries it, and a second stored list would be one more thing
    /// that can fall out of step with the first. Tier raw values never contain
    /// a colon, so the first one is always the separator; an id from a future
    /// shape with no colon degrades to itself, which can then only match
    /// exactly — the safe direction.
    static func quote(inMarkId id: String) -> String {
        guard let separator = id.firstIndex(of: ":") else { return id }
        return String(id[id.index(after: separator)...])
    }

    /// Two quotes are the same passage when either contains the other.
    ///
    /// Short quotes are excluded from the containment half: a three-character
    /// fragment is contained in half the page and would swallow unrelated
    /// marks. Below that length only equality counts.
    static func overlaps(_ left: String, _ right: String) -> Bool {
        if left == right { return true }
        let shortest = min(left.count, right.count)
        guard shortest >= 8 else { return false }
        return left.contains(right) || right.contains(left)
    }
}

extension PageCoverage {
    /// JSON for `CapturedPage.coverageJSON`. Same storage shape as
    /// `ExerciseFilter.storageValue`: one column, no second entity, nothing
    /// that needs a cascade rule.
    public var storageValue: String? {
        let encoder = JSONEncoder()
        encoder.dateEncodingStrategy = .iso8601
        guard let data = try? encoder.encode(self) else { return nil }
        return String(data: data, encoding: .utf8)
    }

    /// Decodes a stored value, or an empty coverage for anything unreadable.
    ///
    /// Never throws and never returns nil: a page whose coverage blob cannot be
    /// read is a page with no findings to show, which is exactly what an empty
    /// value means. Failing louder here would take a whole page detail screen
    /// down over an audit extra.
    public static func fromStorage(_ raw: String?) -> PageCoverage {
        guard let raw, let data = raw.data(using: .utf8) else { return PageCoverage() }
        let decoder = JSONDecoder()
        decoder.dateDecodingStrategy = .iso8601
        return (try? decoder.decode(PageCoverage.self, from: data)) ?? PageCoverage()
    }
}
